wagtail/wagtailsearch/backends/elasticsearch.py

```diff
@@ -307,6 +354,115 @@ def _do_count(self):
         return max(hit_count, 0)
 
 
+class ElasticSearchIndexRebuilder(object):
+    def __init__(self, es, index_name):
+        self.es = es
+        self.index_name = index_name
+
+    def reset_index(self):
+        # Delete old index
+        try:
+            self.es.indices.delete(self.index_name)
+        except NotFoundError:
+            pass
+
+        # Create new index
+        self.es.indices.create(self.index_name, INDEX_SETTINGS)
+
+    def start(self):
+        # Reset the index
+        self.reset_index()
+
+    def add_model(self, model):
+        # Get mapping
+        mapping = ElasticSearchMapping(model)
+
+        # Put mapping
+        self.es.indices.put_mapping(index=self.index_name, doc_type=mapping.get_document_type(), body=mapping.get_mapping())
+
+    def add_items(self, model, obj_list):
+        if not class_is_indexed(model):
+            return
+
+        # Get mapping
+        mapping = ElasticSearchMapping(model)
+        doc_type = mapping.get_document_type()
+
+        # Create list of actions
+        actions = []
+        for obj in obj_list:
+            # Create the action
+            action = {
+                '_index': self.index_name,
+                '_type': doc_type,
+                '_id': mapping.get_document_id(obj),
+            }
+            action.update(mapping.get_document(obj))
+            actions.append(action)
+
+        # Run the actions
+        bulk(self.es, actions)
+
+    def finish(self):
+        # Refresh index
+        self.es.indices.refresh(self.index_name)
+
+
+class ElasticSearchAtomicIndexRebuilder(ElasticSearchIndexRebuilder):
+    def __init__(self, es, alias_name):
+        self.es = es
+        self.alias_name = alias_name
+        self.index_name = alias_name + '_' + get_random_string(7).lower()
+
+    def reset_index(self):
+        # Delete old index using the alias
+        # This should delete both the alias and the index
+        try:
+            self.es.indices.delete(self.alias_name)
+        except NotFoundError:
+            pass
+
+        # Create new index
+        self.es.indices.create(self.index_name, INDEX_SETTINGS)
+
+        # Create a new alias
+        self.es.indices.put_alias(name=self.alias_name, index=self.index_name)
+
+    def start(self):
+        # Create new index
+        self.es.indices.create(self.index_name, INDEX_SETTINGS)
+
+        # Make sure there isn't currently an index that clashes with alias_name
+        # This can happen when the atomic rebuilder is first enabled
+        if not self.es.indices.exists_alias(self.alias_name):
+            try:
+                self.es.indices.delete(self.alias_name)
+
+                # An index was deleted so there couldn't have been an alias there.
+                # Create an alias now so the site search doesn't break
+                self.es.indices.put_alias(name=self.alias_name, index=self.index_name)
```

Comment:
This confused me for a while - I wasn't sure why we were creating the alias at this point (and only when the deletion of a same-named index has succeeded). I think it was the comment that threw me - I'm guessing that was written before the `exists_alias` check was added...

Is there any reason why we can't move this whole block to the `finish` method? That way, we don't need this extra `put_alias`, and it means that the site search can continue to work (with a complete index) during the first rebuild after enabling ATOMIC_REBUILD.